scala/integration/intellij-bsp/src/org/jetbrains/plugins/scala/bsp/server/connection/ScalaConnectionDetailsProvider.scala

```diff
@@ -0,0 +1,127 @@
+      if (connectionFile.isDefined) {
+        cont.resume(true)
+      }
+
+      val hasGeneratedFile = generateSbtConnectionFile(projectPath).isDefined
+      cont.resume(hasGeneratedFile)
+    }, continuation)
+  }
+
+  override def provideNewConnectionDetails(project: Project, bspConnectionDetails: BspConnectionDetails): BspConnectionDetails = {
+    if (bspConnectionDetails != null) {
+      return bspConnectionDetails
+    }
+
+    val projectRootPath = getProjectRootFile(project) match {
+      case Some(projectFile) => projectFile
+      case None =>
+        error("Cannot obtain project path, please reimport the project.")
+        return bspConnectionDetails
+    }
+
+    val connectionFile = getConnectionFile(projectRootPath) match {
+      case None => generateSbtConnectionFile(projectRootPath)
+      case other => other
+    }
+
+    connectionFile match {
+      case None => null
+      case Some(connectionFile) => parseConnectionFile(connectionFile).orNull
+    }
+  }
+
+
+  override def getBuildToolId: BuildToolId = BUILD_TOOL_ID
+
+  private def parseConnectionFile(connectionFile: VirtualFile): Option[BspConnectionDetails] =
+    Option(ParsersKt.parseBspConnectionDetails(connectionFile))
+
+  private def saveProjectRootFile(project: Project, projectRootFile: VirtualFile): Unit =
+    project.getActualComponentManager.getService(classOf[ScalaConnectionDetailsProviderService])
+      .loadState(new ScalaConnectionDetailsProviderMetadata(Option(projectRootFile)))
+
+  private def getProjectRootFile(project: Project): Option[VirtualFile] = {
+    project.getActualComponentManager.getService(classOf[ScalaConnectionDetailsProviderService])
+      .getState
+      .projectRootPath
+  }
+
+  private def getConnectionFile(projectPath: VirtualFile): Option[VirtualFile] = {
+    getChild(projectPath, List.apply(".bsp", "sbt.json"))
+  }
+
+  private def generateSbtConnectionFile(projectPath: VirtualFile): Option[VirtualFile] = {
+    val process = new ProcessBuilder("coursier", "launch", "sbt", "--", "bspConfig")
+      .directory(projectPath.toNioPath.toFile)
+      .start
+
+    if (process.waitFor(2, TimeUnit.MINUTES)) {
+      if (process.exitValue() != 0) {
+        val processInput = process.inputReader.lines.toList.asScala.mkString
+        val processError = process.errorReader.lines.toList.asScala.mkString
+        throw new RuntimeException(s"Sbt bsp file genration with coursier failed. stdout=[$processInput], stderr=[$processError]")
+      }
+
+      getConnectionFile(projectPath)
+    } else {
+      throw new RuntimeException("Sbt bsp file generation timed out")
+    }
+  }
+
+  private def getChild(root: VirtualFile, path: List[String]): Option[VirtualFile] = {
+    val found = path.foldLeft(Option(root)) { (vf, child) =>
+      vf.flatMap(vf => {
+        vf.refresh(false, false)
+        Option(vf.findChild(child))
+      }
+      )
+    }
+
+    found.map(vf => {
+      vf.refresh(false, false)
+      vf
+    })
+  }
```

Comment:
it's not obvious to me what this is supposed to do

Reply:
Given (root: file, eg. /Users/my_folder; yes file is folder ;)) and path ["abc", "xyz", "pan_tadeusz.txt"] we try to traverse down from the file to the given path - so result would be virtual_file pointing to the /Users/my_folder/abc/xyz/pan_tadeusz.txt. We refresh it along the way because some IO may be involved and during that microseconds something in the file system may change.